### PR TITLE
Allow first batch to accept more than one op

### PR DESCRIPTION
**Pull request by aruiz14**
The `Prepare` hook can be a heavy action, so it makes sense to keep
accepting items in the queue while this stage is running.
Now a batch will be closed for any incoming items right be starting
`Process`.

### CODE REVIEW

File: batch.go [hunk 1]
```diff
@@ -52,15 +61,24 @@ func (q *Queue) DoChan(op func() error) <-chan error {
 	go func() {
 		<-done
 		ch <- b.err
+		close(ch)
```
Comment (aruiz14) [inline batch.go:64]: This was missing

Comment (mkmik) [reply batch.go:64]: well, technically you don't have to close all channels. Closing a channel is useful when a channel is used to send an unbounded amount of messages. It basically says: "I hereby signal to all consumers that I'm done producing".

Now, in case you think that closing the channel "just in case" is good practice because if you're not meant to read from it "wouldn't it be better if it's formally closed?", I beg to differ:

Reading from a closed channel is legal and returns the zero value of the channel.
Thus, if you close a channel that is not meant to be ever read again, you might hide a bug.

It's easier to notice that something is wrong if you get a runtime error like:

```
all goroutines are asleep - deadlock!
```

rather than getting a `nil` error and thinking that everything is alright.

That's also why reading from a nil channel blocks in Go. It might be a bit confusing but "closing" a channel is just a signalling feature; it's as if channels had a second data type through which you can carry a "EOF" signal through.

**Comment by mkmik** [comment]
It's getting too complicated to follow; the original code is hard to follow too and adding more logic makes it even worse;
can we start by measuring how often does it happen that a request arrives while a batch is preparing?

File: batch_test.go [hunk 2]
```diff
@@ -98,18 +90,51 @@ func TestError(t *testing.T) {
...
+		n++
+		return &heavyLifting{id: n, waiter: w}
+	}
+	q := batchelor.NewQueue(newJob)
+
+	for j := 1; j < 4; j++ {
+		opsPerBatch := j * 2
+		var processed int
+		errChs := make([]<-chan error, opsPerBatch)
+		for i := 0; i < len(errChs); i++ {
```
Comment (mkmik) [inline batch_test.go:120]: you probably want `cap` instead of `len`: `len` of a channel returns the number of items it contains, not its size

(see https://play.golang.org/p/MlIwECZZqHp)

**Review by mkmik — APPROVED** [review]
anyway, the code seems correct to me

File: batch.go (hunk 1 above)
Comment (aruiz14) [reply batch.go:64]: Understood, thanks for the explanation!
I thought I was preventing a leak.
I'll revert this.

**Comment by aruiz14** [comment]
Let's hold this for now and recover it if we ever notice there is a significant improvement in doing this.
